app/controllers/ileQuery/IleQueryController.scala

```
@@ -0,0 +1,125 @@
+  ileQueryDucrResponsePage: ile_query_ducr_response,
+  ileQueryMucrResponsePage: ile_query_mucr_response
+)(implicit ec: ExecutionContext)
+    extends FrontendController(mcc) with I18nSupport {
+
+  def displayQueryForm(): Action[AnyContent] = authenticate { implicit request =>
+    Ok(ileQueryPage(form))
+  }
+
+  def submitQueryForm(): Action[AnyContent] = authenticate { implicit request =>
+    form
+      .bindFromRequest()
+      .fold(
+        formWithErrors => BadRequest(ileQueryPage(formWithErrors)),
+        validUcr => Redirect(controllers.ileQuery.routes.IleQueryController.submitQuery(validUcr))
+      )
+  }
+
+  def submitQuery(ucr: String): Action[AnyContent] = authenticate.async { implicit request =>
+    def loadingPageResult = Ok(loadingScreenPage()).withHeaders("refresh" -> "5")
+
+    ileQueryRepository.findBySessionIdAndUcr(retrieveSessionId, ucr).flatMap {
+      case Some(query) =>
+        connector.fetchQueryNotifications(query.conversationId, request.providerId).flatMap { response =>
+          response.status match {
+            case OK =>
+              ileQueryRepository.removeByConversationId(query.conversationId).map { _ =>
+                val queryResponse = Json.parse(response.body).as[IleQueryResponse]
+
+                val ducrResult = queryResponse.queriedDucr.map(ducr => Ok(ileQueryDucrResponsePage(ducr)))
+                val mucrResult = queryResponse.queriedMucr.map(mucr => Ok(ileQueryMucrResponsePage(mucr)))
+
+                ducrResult.orElse(mucrResult).getOrElse(loadingPageResult)
+              }
+            case NO_CONTENT =>
+              Future.successful(loadingPageResult)
+            case GATEWAY_TIMEOUT =>
+              ileQueryRepository.removeByConversationId(query.conversationId).map { _ =>
+                InternalServerError(errorHandler.standardErrorTemplate())
+              }
+            case _ => Future.successful(BadRequest(errorHandler.standardErrorTemplate()))
```

I don't think this a case for `BadRequest`. To me it is more like `InternalServerError`. So maybe just throw an exception in this case? Or "let it crash" by having `MatchException`

---

Will change to `InternalServerError`

---

Changed